txpipe/twopoint_fourier.py

```diff
@@ -348,137 +358,177 @@ def select_calculations(self, nbins_source, nbins_lens):
-            workspace = w00
             field_i = f_d[i]
             field_j = f_d[j]
-            cl_guess = [theory]
-            results_to_use = [(0, Cdd)]
+            results_to_use = [(0, Cdd, pixwin_t**2)]
 
         elif k == SHEAR_POS:
-            workspace = w02
             field_i = f_wl[i]
             field_j = f_d[j]
-            cl_guess = [theory, np.zeros_like(theory)]
-            results_to_use = [(0, CdE), (1, CdB)]
+            results_to_use = [(0, CdE, pixwin_t*pixwin_p), (1, CdB, pixwin_t*pixwin_p)]
+
+        if pixel_scheme.name == 'healpix':
+            workspace = nmt.NmtWorkspace()
+        elif pixel_scheme.name == 'gnomonic':
+            workspace = nmt.NmtWorkspaceFlat()
+        else:
+            raise ValueError(f"No NaMaster workspace for pixel scheme {pixel_scheme.name}")
+
+        # Compute mode-coupling matrix
+        workspace.compute_coupling_matrix(field_i, field_j, ell_bins)
 
-        cl_noise = self.compute_noise(i,j,k,ell_bins,workspace,noise)
+        # Get the coupled noise C_ell values to give to the master algorithm
+        cl_noise = self.compute_noise(i, j, k, ell_bins, workspace, tomo_info)
 
+        # Run the master algorithm
         if pixel_scheme.name == 'healpix':
-            c = pymaster.compute_full_master(field_i, field_j, ell_bins,
+            c = nmt.compute_full_master(field_i, field_j, ell_bins,
                 cl_noise=cl_noise, cl_guess=cl_guess, workspace=workspace)
         elif pixel_scheme.name == 'gnomonic':
-            c = pymaster.compute_full_master_flat(field_i, field_j, ell_bins,
-                cl_noise=cl_noise, cl_guess=cl_guess, ells_guess=cl_theory['ell'],
+            c = nmt.compute_full_master_flat(field_i, field_j, ell_bins,
+                cl_noise=cl_noise, cl_guess=cl_guess, ells_guess=ell_guess,
                 workspace=workspace)
 
-        for index, name in results_to_use:
-            self.results.append(Measurement(name, ls, c[index], win, i, j))
+        # Save all the results, skipping things we don't want like EB modes
+        for index, name, pixwin in results_to_use:
+            self.results.append(Measurement(name, ls, c[index] / pixwin, win, i, j))
 
 
-    def compute_noise(self, i, j, k, ell_bins, w, noise):
-        # No noise contribution from cross-correlations
+    def compute_noise(self, i, j, k, ell_bins, w, tomo_info):
+        # No noise contribution in cross-correlations
         if (i!=j) or (k==SHEAR_POS):
             return None
-        # print("Setting noise to zero")
-        # return None
+
         # We loaded in sigma_e and the densities
-        # earlier on and put them in the noise dictionary
-        noise_level = noise[(i,k)]
+        # earlier on, and put them in the tomo_info dictionary
+        if k==SHEAR_SHEAR:
+            noise_level = tomo_info['sigma_e'][i]**2 / tomo_info['n_eff_steradian'][i]
+        else:
+            noise_level = 1.0 / tomo_info['n_lens_steradian'][i]
```

Comment:
Same here. You should make this `np.mean(mask) / tomo_info['n_lens_steradian']` (assuming `tomo_info['n_lens_steradian']` was computed as `np.sum(number_map * mask) / np.sum(mask)` as I mentioned above).